### PR TITLE
Track using violations

### PR DESCRIPTION
**Pull request by jridgewell**
There's no way we can visit every identifier. Instead, figure out which
constant violations hamper the minification.

Oh, and some test cases. Those should probably be cleaned up. :grin:

### CODE REVIEW

**Comment by jridgewell** [comment]
Also, the jQuery benchmark runs some ~40ms faster, react ~30ms.

File: packages/babel-plugin-transform-remove-undefined/src/index.js [hunk 1]
```diff
@@ -65,12 +46,46 @@ module.exports = function({ types: t }) {
...
+              for (let func = getFunctionParent(v, scopeParent); func; func = getFunctionParent(func, scopeParent)) {
+                const id = func.node.id;
+                const binding = id && scope.getBinding(id.name);
+
+                if (!binding) {
+                  continue;
+                }
+
+                const funcViolation = binding.referencePaths.some((p) => {
+                  return p.node.start < start;
```
Comment (shinew) [inline packages/babel-plugin-transform-remove-undefined/src/index.js]: Would this fail for references through multiple levels of indirection?

e.g.

``` javascript
function foo() {
  bar();
  var a = undefined;
  console.log(a);
  function bar() {
    g();
  }
  function g() {
    a = 4;
  }
}
```

We cannot remove the `= undefined` in this case. As I understand this, it only checks for 1 level of referencing.

File: packages/babel-plugin-transform-remove-undefined/src/index.js [hunk 2]
```diff
@@ -65,12 +46,46 @@ module.exports = function({ types: t }) {
...
-          }
+          const { node, scope } = path;
           for (const declarator of path.get("declarations")) {
-            if (areAllBindingsNotSeen(declarator, names, t)) {
+            const binding = scope.getBinding(declarator.node.id.name);
+            if (!binding) {
+              continue;
+            }
+
+            const { start } = node;
```
Comment (shinew) [inline packages/babel-plugin-transform-remove-undefined/src/index.js:75]: I think the `start` and `end` attributes won't work for nodes that are runtime-generated: https://astexplorer.net/#/Pl6ASeZRlL. This is still probably better than any other way of determining relative ordering though, if we add some existence checks for these attributes for safety.

**Comment by shinew** [comment]
This looks like a better approach, if we can resolve the edge cases in my comments (above). @kangax @boopathi thoughts?

File: packages/babel-plugin-transform-remove-undefined/src/index.js (hunk 1 above)
Comment (jridgewell) [reply packages/babel-plugin-transform-remove-undefined/src/index.js]: That'll be easy enough to add.

File: packages/babel-plugin-transform-remove-undefined/src/index.js (hunk 2 above)
Comment (jridgewell) [reply packages/babel-plugin-transform-remove-undefined/src/index.js:75]: Ah, I thought doing a `#replaceWith` auto-inherited the location. Looks like you have to `t.inherits(newNode, oldNode)` in order to get that.

**Comment by shinew** [comment]
@jridgewell I'll merge this once the multi-layer referencing is ready.

File: packages/babel-plugin-transform-remove-undefined/src/index.js (hunk 1 above)
Comment (jridgewell) [reply packages/babel-plugin-transform-remove-undefined/src/index.js]: Done, but it's unoptimized.

**Comment by shinew** [comment]
Nice!

**Pull request merged**